compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -391,22 +395,53 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                                    self.ty_to_value_string(actual.peel_refs())
-                                }
-                                _ => self.ty_to_value_string(ty.peel_refs()),
+                    let ty_str = if let Some(CandidateSource::Impl(impl_did)) =
+                        static_sources.get(0)
+                    {
+                        // When the "method" is resolved through dereferencing, we really want the
+                        // original type that has the associated function for accurate suggestions.
+                        // (#61411)
+                        let ty = tcx.at(span).type_of(*impl_did);
+                        match (&ty.peel_refs().kind(), &actual.peel_refs().kind()) {
+                            (ty::Adt(def, _), ty::Adt(def_actual, substs)) if def == def_actual => {
+                                // If there are any inferred arguments, (`{integer}`), we should replace
+                                // them with underscores to allow the compiler to infer them
+                                let substs = substs
+                                    .into_iter()
+                                    .filter(|arg| !arg.is_suggestable(tcx, true))
+                                    .map(|arg| match arg.unpack() {
+                                        GenericArgKind::Lifetime(_) => self
+                                            .next_region_var(RegionVariableOrigin::MiscVariable(
+                                                rustc_span::DUMMY_SP,
+                                            ))
+                                            .into(),
+                                        GenericArgKind::Type(_) => self
+                                            .next_ty_var(TypeVariableOrigin {
+                                                span: rustc_span::DUMMY_SP,
+                                                kind: TypeVariableOriginKind::MiscVariable,
+                                            })
+                                            .into(),
+                                        GenericArgKind::Const(arg) => self
+                                            .next_const_var(
+                                                arg.ty(),
+                                                ConstVariableOrigin {
+                                                    span: rustc_span::DUMMY_SP,
+                                                    kind: ConstVariableOriginKind::MiscVariable,
+                                                },
+                                            )
+                                            .into(),
+                                    })
+                                    .collect::<Vec<_>>();
+                                format!(
+                                    "{}",
+                                    ty::Instance::new(def_actual.did(), tcx.intern_substs(&substs))
```

Comment:
I don't think using `Instance` here is right. Can you instead try using something like `tcx.def_path` or [`tcx.def_path_with_substs`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.def_path_str_with_substs)?

Reply:
`def_path_with_substs` requires that `substs` is borrowed for `'tcx`. Since it's local to the match arm it doesn't live long enough for this